From a043d160dd8d1f5b92cadf1686ce9f3df55b9599 Mon Sep 17 00:00:00 2001
From: kargl <kargl@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Thu, 16 Jul 2015 19:03:05 +0000
Subject: [PATCH] 2015-07-16  Steven G. Kargl   <kargl@gcc.gnu.org>

	PR fortran/66545
	* primary.c (match_sym_complex_part): Do not dereference NULL pointer.

2015-07-16  Steven G. Kargl   <kargl@gcc.gnu.org>

	PR fortran/66545
	* gfortran.dg/pr66545_1.f90: New test.
	* gfortran.dg/pr66545_2.f90: New test.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-5-branch@225901 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/fortran/ChangeLog                   |  5 +++++
 gcc/fortran/primary.c                   |  3 +++
 gcc/testsuite/ChangeLog                 |  6 ++++++
 gcc/testsuite/gfortran.dg/pr66545_1.f90 | 17 +++++++++++++++++
 gcc/testsuite/gfortran.dg/pr66545_2.f90 | 23 +++++++++++++++++++++++
 5 files changed, 54 insertions(+)
 create mode 100644 gcc/testsuite/gfortran.dg/pr66545_1.f90
 create mode 100644 gcc/testsuite/gfortran.dg/pr66545_2.f90

diff --git a/gcc/fortran/ChangeLog b/gcc/fortran/ChangeLog
index 816e0b2..1133082 100644
--- a/gcc/fortran/ChangeLog
+++ b/gcc/fortran/ChangeLog
@@ -1,3 +1,8 @@
+2015-07-16  Steven G. Kargl   <kargl@gcc.gnu.org>
+
+	PR fortran/66545
+	* primary.c (match_sym_complex_part): Do not dereference NULL pointer.
+
 2015-07-16  Steven G. Kargl  <kargl@gcc.gnu.org>
 
 	PR fortran/56520
diff --git a/gcc/fortran/primary.c b/gcc/fortran/primary.c
index e9ced7e..59819ee 100644
--- a/gcc/fortran/primary.c
+++ b/gcc/fortran/primary.c
@@ -1202,6 +1202,9 @@ match_sym_complex_part (gfc_expr **result)
       return MATCH_ERROR;
     }
 
+  if (!sym->value)
+    goto error;
+
   if (!gfc_numeric_ts (&sym->value->ts))
     {
       gfc_error ("Numeric PARAMETER required in complex constant at %C");
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index a55da11..0b5de0e 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,9 @@
+2015-07-16  Steven G. Kargl   <kargl@gcc.gnu.org>
+
+	PR fortran/66545
+	* gfortran.dg/pr66545_1.f90: New test.
+	* gfortran.dg/pr66545_2.f90: New test.
+
 2015-07-16  Steven G. Kargl  <kargl@gcc.gnu.org>
 
 	PR fortran/56520
diff --git a/gcc/testsuite/gfortran.dg/pr66545_1.f90 b/gcc/testsuite/gfortran.dg/pr66545_1.f90
new file mode 100644
index 0000000..7daa800
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr66545_1.f90
@@ -0,0 +1,17 @@
+! { dg-do compile }
+! { dg-options "-Wall" }
+! PR fortran/66545
+!
+subroutine p
+   complex, parameter :: c1 = (c1) ! { dg-error "before its definition" }
+   complex, parameter :: c2 = c2   ! { dg-error "before its definition" }
+   complex :: c3 = (c3)            ! { dg-error "has not been declared or is a variable" }
+   complex :: c4 = c4              ! { dg-error "has not been declared or is a variable" }
+end subroutine p
+
+subroutine q
+   real, parameter :: r1 = (r1)  ! { dg-error "before its definition" }
+   real, parameter :: r2 = r2    ! { dg-error "before its definition" }
+   real :: r3 = (r3)             ! { dg-error "has not been declared or is a variable" }
+   real :: r4 = r4               ! { dg-error "has not been declared or is a variable" }
+end subroutine q
diff --git a/gcc/testsuite/gfortran.dg/pr66545_2.f90 b/gcc/testsuite/gfortran.dg/pr66545_2.f90
new file mode 100644
index 0000000..e15d8ba
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr66545_2.f90
@@ -0,0 +1,23 @@
+! { dg-do compile }
+! { dg-options "-Wuninitialized" }
+! PR fortran/66545
+!
+program foo
+   implicit none
+   call p1
+   call q1
+end program foo
+
+subroutine p1
+   complex :: c5
+   complex :: c6
+   c5 = (c5)      ! { dg-warning "used uninitialized in this" }
+   c6 = c6        ! { dg-warning "used uninitialized in this" }
+end subroutine p1
+
+subroutine q1
+   real :: r5
+   real :: r6
+   r5 = (r5)   ! { dg-warning "used uninitialized in this" }
+   r6 = r6     ! { dg-warning "used uninitialized in this" }
+end subroutine q1
-- 
1.9.4

